firedrake/slate/slac/utils.py

```diff
@@ -137,6 +136,37 @@ def visit_Symbol(self, o, *args, **kwargs):
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
 
 
+def topological_sort(exprs):
+    """Returns a topologically sorted list of Slate expressions.
+    The list is generated by examining the dependencies of each
+    expression and moving nodes that have all its dependencies
+    resolved. A node is resolved once all the nodes in its dependencies
+    have been moved into the sorted list.
+
+    :arg exprs: An iterable of Slate nodes.
+    """
+    # Don't include the expr itself in its dependencies
+    dependency_graph = {expr: list(set(traverse_dags([expr])) - {expr})
+                        for expr in exprs}
+
+    sorted_exprs = []
+    while dependency_graph:
+        acyclic = False
+        for tensor, dependencies in list(dependency_graph.items()):
+            for exp in dependencies:
+                if exp in dependency_graph:
+                    break
+            else:
+                acyclic = True
+                del dependency_graph[tensor]
+                sorted_exprs.append(tensor)
+
+        if not acyclic:
+            raise RuntimeError("A cyclic dependency found.")
+
+    return sorted_exprs
```

Comment:
How about:
```python
def topo_sort(exprs, depgraph):
    depgraph = OrderedDict(((expr, set(traverse_dags([expr])) - {expr}) for expr in exprs))
    schedule = []
    while depgraph:
        for expr in depgraph:
            if not depgraph[expr].intersection(depgraph):
                schedule.append(expr)
                break
        else:
            raise ValueError("Expression graph is cyclic")
        depgraph.pop(expr)
    return schedule
```

Reply:
Thanks, I like this version more. I'll update.